crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -1166,18 +1179,138 @@ impl<'db> TypeInferenceBuilder<'db> {
+                                let remaining = elts.len() - 2;
+                                let end_index =
+                                    tuple_ty.len(builder.db) - starred_index - remaining;
+                                // SAFETY: Safe because of the length check above.
+                                let _starred_element_types = tuple_ty
+                                    .elements(builder.db)
+                                    .get(starred_index..=end_index)
+                                    .unwrap();
+                                // TODO: Combine the types into a list type.
+                                // combine_types(starred_element_types);
+                                element_types.push(Type::Todo);
+
+                                element_types.extend_from_slice(
+                                    // SAFETY: Safe because of the length check above.
+                                    tuple_ty.elements(builder.db).get(end_index + 1..).unwrap(),
+                                );
+                                Cow::Owned(element_types)
+                            } else if tuple_ty.len(builder.db) == elts.len() - 1 {
+                                let mut element_types = tuple_ty.elements(builder.db).to_vec();
+                                element_types.insert(starred_index, Type::Todo);
+                                Cow::Owned(element_types)
+                            } else {
+                                Cow::Borrowed(tuple_ty.elements(builder.db).as_ref())
+                            }
+                        } else {
+                            Cow::Borrowed(tuple_ty.elements(builder.db).as_ref())
+                        };
+
+                        for (index, element) in elts.iter().enumerate() {
+                            if let Some(ty) = inner(
+                                builder,
+                                element,
+                                element_types.get(index).copied().unwrap_or(Type::Unknown),
+                                variable,
+                            ) {
+                                return Some(ty);
+                            }
+                        }
+                    }
+                    Type::StringLiteral(string_literal_ty) => {
+                        for (index, element) in elts.iter().enumerate() {
+                            let string_length = string_literal_ty.len(builder.db);
+                            let element_ty = if index < string_length {
+                                Type::LiteralString
+                            } else {
+                                Type::Unknown
+                            };
+                            if let Some(ty) = inner(builder, element, element_ty, variable) {
+                                return Some(ty);
+                            }
+                        }
+                    }
```

Comment:
It seems like to assign Unknown correctly here (and in the future to correctly issue diagnostics for mismatched unpacking) we also need to account for starred element, if present?

Alternatively, we could just generalize the tuple RHS case above, so for a string RHS we get a slice of one-character StringLiteral types and otherwise all the logic is identical? That would avoid needing to implement the starred index handling twice.

I don't think it's necessary that we precisely infer literal string types from unpacking a literal string, only suggesting that in case it is just as easy as handling starred element explicitly here?

Reply:
Hmm, it might be useful to generalize it. We can convert the `StringLiteral` type into a `TupleType` containing `n` number of either `LiteralString` or `StringLiteral` (with the corresponding character) where `n` is the length of the containing string.

Reply:
I've addressed this feedback by splitting the `StringLiteral` type into a tuple type containing equal amount of `LiteralString` and delegating it back to the `inner` function for correct handling of starred expression.